lib/kube/proxy/forwarder.go

```diff
@@ -652,16 +669,78 @@ func (f *Forwarder) setupForwardingHeaders(ctx *authContext, sess *clusterSessio
 	req.Header.Add("X-Forwarded-Path", req.URL.Path)
 
+	return nil
+}
+
+// setupImpersonationHeaders sets up Impersonate-User and Impersonate-Group headers
+func setupImpersonationHeaders(log log.FieldLogger, ctx *authContext, headers http.Header, bearerToken string) error {
+	var impersonateUser string
+	var impersonateGroups []string
+	for header, values := range headers {
+		if !strings.HasPrefix(header, "Impersonate-") {
+			continue
+		}
+		switch header {
+		case ImpersonateUserHeader:
+			if impersonateUser != "" {
+				return trace.AccessDenied("%v, user already specified to %q", ImpersonationRequestDeniedMessage, impersonateUser)
+			}
+			if len(values) == 0 || len(values) > 1 {
+				return trace.AccessDenied("%v, invalid user header %q", ImpersonationRequestDeniedMessage, values)
+			}
+			impersonateUser = values[0]
+			if _, ok := ctx.kubeUsers[impersonateUser]; !ok {
+				return trace.AccessDenied("%v, user header %q is not allowed in roles", ImpersonationRequestDeniedMessage, impersonateUser)
+			}
+		case ImpersonateGroupHeader:
+			for _, group := range values {
+				if _, ok := ctx.kubeGroups[group]; !ok {
+					return trace.AccessDenied("%v, group header %q value is not allowed in roles", ImpersonationRequestDeniedMessage, group)
+				}
+				impersonateGroups = append(impersonateGroups, group)
+			}
+		default:
+			return trace.AccessDenied("%v, unsupported impersonation header %q", ImpersonationRequestDeniedMessage, header)
+		}
+	}
+
+	impersonateGroups = utils.Deduplicate(impersonateGroups)
+
+	// By default, if no kubernetes_users is set (which will be a majority),
+	// user will impersonate themselves, which is the backwards-compatible behavior.
+	//
+	// As long as at least one `kubernetes_users` is set, the forwarder will start
+	// limiting the list of users allowed by the client to impersonate.
+	//
+	// If the users' role set does not include actual user name, it will be rejected,
+	// otherwise there will be no way to exclude the user from the list).
+	//
+	// If the `kuberentes_users` role set includes only one user
+	// (quite frequently that's the real intent), teleport will default to it,
+	// otherwise it will refuse to select.
+	//
+	// This will enable the use case when `kubernetes_users` has just one field to
+	// link the user identity with the IAM role, for example `IAM#{{external.email}}`
+	//
+	if impersonateUser == "" && len(ctx.kubeUsers) == 1 {
+		for user := range ctx.kubeUsers {
+			impersonateUser = user
+			break
+		}
+	}
+	if len(impersonateGroups) == 0 {
+		for group := range ctx.kubeGroups {
+			impersonateGroups = append(impersonateGroups, group)
+		}
+	}
+
 	if !ctx.cluster.isRemote {
-		req.Header.Add("Impersonate-User", ctx.User.GetName())
-		f.Debugf("Impersonate User: %v", ctx.User)
-		for _, group := range ctx.kubeGroups {
-			req.Header.Add("Impersonate-Group", group)
-			f.Debugf("Impersonate Group: %v", group)
+		headers.Add("Impersonate-User", impersonateUser)
```

Comment:
It appears that `impersonateUser` may be the empty string (`""`) if `len(ctx.kubeUsers) != 1`.  Is that intended behavior?

Reply:
@fspmarshall thanks for outstanding review detail. I think this is undefined behavior, however I restricted our approach to rejecting the empty impersonating header and in case of multiple choice we will refuse to make a selection, asking user to select a user, see my changes above.